### PR TITLE
`azurerm_kusto_cluster` - Support `None pattern` for `virtual_network_configuration` block

### PR DESCRIPTION
**Pull request by liuwuliuyun**
`azurerm_kusto_cluster`
- Split create and update function in this resource.
- Support `None pattern` for `virtual_network_configuration` block.
- [None pattern coding guidline](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/topics/schema-design-considerations.md#the-none-value-or-similar)
- Removing `forceNew` from `virtual_network_configuration` block.
- Adding new test example.
- Fixes https://github.com/hashicorp/terraform-provider-azurerm/issues/24739
Testing evidence:
```
GOROOT=C:\Program Files\Go #gosetup
GOPATH=C:\Users\yunliu1\go #gosetup
"C:\Program Files\Go\bin\go.exe" test -c -o C:\Users\yunliu1\AppData\Local\JetBrains\GoLand2023.3\tmp\GoLand\___TestAccKustoCluster_vnet_in_github_com_hashicorp_terraform_provider_azurerm_internal_services_kusto.test.exe github.com/hashicorp/terraform-provider-azurerm/internal/services/kusto #gosetup
"C:\Program Files\Go\bin\go.exe" tool test2json -t C:\Users\yunliu1\AppData\Local\JetBrains\GoLand2023.3\tmp\GoLand\___TestAccKustoCluster_vnet_in_github_com_hashicorp_terraform_provider_azurerm_internal_services_kusto.test.exe -test.v -test.paniconexit0 -test.run ^\QTestAccKustoCluster_vnet\E$ #gosetup
=== RUN   TestAccKustoCluster_vnet
=== PAUSE TestAccKustoCluster_vnet
=== CONT  TestAccKustoCluster_vnet
--- PASS: TestAccKustoCluster_vnet (2300.45s)
PASS


Process finished with the exit code 0
```

### CODE REVIEW

File: internal/services/kusto/kusto_cluster_resource.go [hunk 1]
```diff
@@ -387,20 +381,188 @@ func resourceKustoClusterCreateUpdate(d *pluginsdk.ResourceData, meta interface{
...
+
+	if d.HasChange("virtual_network_configuration") {
+		if v, ok := d.GetOk("virtual_network_configuration"); ok {
+			if vnetConfig := expandKustoClusterVNET(v.([]interface{})); vnetConfig != nil {
+				props.VirtualNetworkConfiguration = vnetConfig
+			}
+		} else {
+			props.VirtualNetworkConfiguration.State = pointer.To(clusters.VnetStateDisabled)
+			props.PublicNetworkAccess = pointer.To(clusters.PublicNetworkAccessDisabled)
+		}
```
Comment (WodansSon) [inline internal/services/kusto/kusto_cluster_resource.go:541]: I thought `PublicNetworkAccess` and `VirtualNetworkConfiguration.State` were independent of each other and could be toggled seperately?
```suggestion
		} else {
			// 'State' is hardcoded to 'Disabled' for the 'None' pattern.
			// If the vNet block is present it is enabled, if the vNet block is removed it is disabled.
			props.VirtualNetworkConfiguration.State = pointer.To(clusters.VnetStateDisabled)
		}
```

File: internal/services/kusto/kusto_cluster_resource.go [hunk 2]
```diff
@@ -579,6 +741,7 @@ func expandKustoClusterVNET(input []interface{}) *clusters.VirtualNetworkConfigu
 		SubnetId:                 subnetID,
 		EnginePublicIPId:         enginePublicIPID,
 		DataManagementPublicIPId: dataManagementPublicIPID,
+		State:                    pointer.To(clusters.VnetStateEnabled),
```
Comment (WodansSon) [inline internal/services/kusto/kusto_cluster_resource.go:744]: Can we add some comments here to explain why we have this logic here?
```suggestion
		// 'State' is hardcoded to 'Enabled' for the 'None' pattern.
		// If the vNet block is present it is enabled, if the vNet block is removed it is disabled.
		SubnetId:                 subnetID,
		EnginePublicIPId:         enginePublicIPID,
		DataManagementPublicIPId: dataManagementPublicIPID,
		State:                    pointer.To(clusters.VnetStateEnabled),
```

File: internal/services/kusto/kusto_cluster_resource.go [hunk 3]
```diff
@@ -145,23 +145,25 @@ func resourceKustoCluster() *pluginsdk.Resource {
 			"virtual_network_configuration": {
```
Comment (WodansSon) [inline internal/services/kusto/kusto_cluster_resource.go:145]: Does this block need to be computed optional?

File: internal/services/kusto/kusto_cluster_resource_test.go [hunk 4]
```diff
@@ -241,6 +241,15 @@ func TestAccKustoCluster_vnet(t *testing.T) {
...
 		data.ImportStep(),
+		{
+			Config: r.vnetRemoved(data),
+			Check: acceptance.ComposeTestCheckFunc(
+				check.That(data.ResourceName).ExistsInAzure(r),
+				check.That(data.ResourceName).Key("virtual_network_configuration.#").HasValue("0"),
+				check.That(data.ResourceName).Key("public_network_access_enabled").HasValue("false"),
+			),
+		},
+		data.ImportStep(),
```
Comment (WodansSon) [inline internal/services/kusto/kusto_cluster_resource_test.go:252]: Can we add an additional test step where we add the vNet block back into the config?
```suggestion
		{
			Config: r.vnetRemoved(data),
			Check: acceptance.ComposeTestCheckFunc(
				check.That(data.ResourceName).ExistsInAzure(r),
				check.That(data.ResourceName).Key("virtual_network_configuration.#").HasValue("0"),
			),
		},
		data.ImportStep(),
		{
			Config: r.vnet(data),
			Check: acceptance.ComposeTestCheckFunc(
				check.That(data.ResourceName).ExistsInAzure(r),
				check.That(data.ResourceName).Key("virtual_network_configuration.#").HasValue("1"),
				check.That(data.ResourceName).Key("virtual_network_configuration.0.subnet_id").Exists(),
				check.That(data.ResourceName).Key("virtual_network_configuration.0.engine_public_ip_id").Exists(),
				check.That(data.ResourceName).Key("virtual_network_configuration.0.data_management_public_ip_id").Exists(),
			),
		},
		data.ImportStep(),
```

**Review by WodansSon — CHANGES_REQUESTED** [review]
@liuwuliuyun, this is looking really good so far. I left a few comments which I believe should be addressed. Once that is done this LGTM! 🚀

File: internal/services/kusto/kusto_cluster_resource.go (hunk 3 above)
Comment (liuwuliuyun) [reply internal/services/kusto/kusto_cluster_resource.go:145]: I thought Hashicorp doesn't allow any property marked as computed optional and user has to manually ignore changes for Computed Optional property. Should we added this to the block?

File: internal/services/kusto/kusto_cluster_resource.go (hunk 1 above)
Comment (liuwuliuyun) [reply internal/services/kusto/kusto_cluster_resource.go]: Setting the state to false will automatically set public network access to disabled. In this way, users need to manually set public network access property and let me write that down in doc.

File: website/docs/r/kusto_cluster.html.markdown [hunk 5]
```diff
@@ -68,7 +68,9 @@ The following arguments are supported:
 
 * `purge_enabled` - (Optional) Specifies if the purge operations are enabled.
 
-* `virtual_network_configuration` - (Optional) A `virtual_network_configuration` block as defined below. Changing this forces a new resource to be created.
+* `virtual_network_configuration` - (Optional) A `virtual_network_configuration` block as defined below.
+
+~> **NOTE:** Currently removing `virtual_network_configuration` sets the `virtual_network_configuration` to `Disabled` state. But any changes to `virtual_network_configuration` in `Disabled` state forces a new resource to be created.
```
Comment (katbyte) [inline website/docs/r/kusto_cluster.html.markdown:73]: This seems togo against how terraform is intended to work. i think we should just do a 2nd update to set what is in config if the api changes it?

**Review by katbyte — CHANGES_REQUESTED** [review]

File: website/docs/r/kusto_cluster.html.markdown (hunk 5 above)
Comment (liuwuliuyun) [reply website/docs/r/kusto_cluster.html.markdown:73]: The simplest way to solve this is adding a `state` property to this block like https://github.com/hashicorp/terraform-provider-azurerm/pull/23912?

Comment (katbyte) [reply website/docs/r/kusto_cluster.html.markdown:73]: sorry I left this comment in the wrong place about the wrong thing

**Comment by liuwuliuyun** [comment]
Pending for a re-review from parter team

**Review by katbyte — APPROVED** [review]
LGTM 🌵

**Pull request merged**